File: app/ferramentas/extratus/web/routes/relatorios_robo.py
import zipfile
from datetime import datetime
from io import BytesIO
from pathlib import Path
from urllib.parse import quote

from fastapi import APIRouter, Depends, Form, HTTPException, Request
from fastapi.responses import FileResponse, RedirectResponse, Response

from app.ferramentas.extratus.db.checagem_fila import resolver_solicitantes
from app.ferramentas.extratus.db.jobs import excluir_job, listar_jobs_robo, marcar_notificacao_resolvida_robo, obter_job
from app.ferramentas.extratus.web.rotulos import (
    ABA_RELATORIOS_ROBO,
    FERRAMENTA_SLUG,
    contagem_nav_conferencias_fila,
    contagem_nav_conferencias_manual,
    contagem_nav_relatorios,
    contagem_nav_relatorios_robo,
    rotulo_erro,
    rotulo_status,
)
from app.plataforma.db.models import Usuario
from app.plataforma.db.usuarios import listar_todos_usuarios, marcar_aba_vista
from app.plataforma.web.auth import exigir_acesso_ferramenta, exigir_admin
from app.plataforma.web.templates_util import criar_templates


# "Relatórios do Robô" — repositório universal do que o ROBÔ já
# processou (pronto, em revisão ou com erro), separado da tela
# "Relatórios URGENTES" (só manuais) desde 2026-08-08. Henrique,
# 2026-08-11: acesso de VER esse acervo é do mesmo nível que "Relatórios
# URGENTES" (qualquer um
# com a ferramenta liberada, é acervo do escritório) — não precisa mais
# de acesso à Fila do Robô pra isso; Fila do Robô continua restrita,
# essa é só a permissão de alimentar/operar o Robô, não de ver o que
# ele já produziu.
router = APIRouter(dependencies=[Depends(exigir_acesso_ferramenta("extratus"))])

TEMPLATES_DIR = Path(__file__).resolve().parent.parent / "templates"
PLATAFORMA_TEMPLATES_DIR = (
    Path(__file__).resolve().parents[4] / "plataforma" / "web" / "templates"
)
templates = criar_templates([TEMPLATES_DIR, PLATAFORMA_TEMPLATES_DIR])
templates.env.filters["rotulo_status"] = rotulo_status
templates.env.filters["rotulo_erro"] = rotulo_erro
templates.env.globals["contagem_nav_conferencias_manual"] = contagem_nav_conferencias_manual
templates.env.globals["contagem_nav_conferencias_fila"] = contagem_nav_conferencias_fila
templates.env.globals["contagem_nav_relatorios"] = contagem_nav_relatorios
templates.env.globals["contagem_nav_relatorios_robo"] = contagem_nav_relatorios_robo


@router.get("/relatorios-robo")
def pagina_relatorios_robo(
    request: Request,
    usuario: Usuario = Depends(exigir_acesso_ferramenta("extratus")),
    processo: str | None = None,
    erro: str | None = None,
    sucesso: str | None = None,
):
    jobs = listar_jobs_robo()

    # Henrique, diretoria, 2026-08-27: a diretoria perguntou "o
    # coordenador fulano colocou os processos que pedi no robô?" e não
    # dava pra responder — "Robô automático" sozinho não diz QUEM pediu.
    # `job.solicitante_id` já vem carregado desde o upload pra relatório
    # NOVO (ver ChecagemFila.solicitante_id/checagem_fila.
    # registrar_pendente). Relatório de ANTES dessa coluna existir não
    # tem como ter isso preenchido — `resolver_solicitantes` cai pra
    # dedução por nome+horário nesses casos (Henrique, mesmo dia: "os
    # relatórios que já estavam prontos agora estão como não
    # identificado... manter aquela solução de antes como fallback").
    # Mesmo formato `nomes_por_id` de relatorios_manuais.py, pra
    # reaproveitar o mesmo bloco visual (.relatorio-solicitante) já
    # existente.
    nomes_por_id = {u.id: u.nome for u in listar_todos_usuarios()}
    solicitante_por_job_id = resolver_solicitantes(jobs)
    ids_solicitantes_reais = {sid for sid in solicitante_por_job_id.values() if sid}

    # Só os solicitantes que de fato aparecem na lista atual — dropdown
    # do filtro "Solicitado por", ordenado por nome (não a base de
    # usuários inteira, a maioria nunca mandou nada pro Robô). Henrique,
    # 2026-09-02: o PRÓPRIO usuário NÃO entra aqui à força só pra ter uma
    # opção — se ele nunca pediu nada, não faz sentido oferecer "ver só
    # os meus" como opção selecionável (ficaria sempre vazio); ver
    # `sem_solicitacoes_proprias` abaixo, que cobre esse caso à parte.
    solicitantes_disponiveis = sorted(
        (
            {"id": usuario_id, "nome": nomes_por_id.get(usuario_id, f"Usuário #{usuario_id}")}
            for usuario_id in ids_solicitantes_reais
        ),
        key=lambda item: item["nome"].lower(),
    )

    # Henrique, 2026-09-02: não-admin já abre a tela filtrado em "só os
    # meus" (ver filtro-solicitante no template + relatorios_robo.js) —
    # `solicitante_padrao` é só o valor inicial, a pessoa pode trocar
    # livremente pra "Todos" depois. Quando ela NUNCA pediu nada ainda,
    # não tem um valor real pra usar como padrão (ela nem aparece em
    # `solicitantes_disponiveis`) — nesse caso a tela mostra uma mensagem
    # dedicada no lugar da lista (ver template) em vez de aplicar um
    # filtro que corresponderia a uma opção inexistente no dropdown.
    solicitante_padrao = None
    sem_solicitacoes_proprias = False
    if not usuario.eh_admin:
        if usuario.id in ids_solicitantes_reais:
            solicitante_padrao = usuario.id
        else:
            sem_solicitacoes_proprias = True

    # Renderiza PRIMEIRO, marca como visto DEPOIS — mesmo motivo de
    # gerar_relatorio.py (senão o badge dessa própria visita nunca apareceria).
    resposta = templates.TemplateResponse(
        request,
        "relatorios_robo.html",
        {
            "usuario": usuario,
            "jobs": jobs,
            "nomes_por_id": nomes_por_id,
            "solicitante_por_job_id": solicitante_por_job_id,
            "solicitantes_disponiveis": solicitantes_disponiveis,
            "solicitante_padrao": solicitante_padrao,
            "sem_solicitacoes_proprias": sem_solicitacoes_proprias,
            # Deep-link vindo do botão "Ir ao relatório" (Conferências
            # manuais, web/routes/gerar_relatorio.py, quando o duplicado é do
            # Robô) — pré-preenche a busca, troca pra aba certa
            # (Sucesso/Revisão/Erro) e dá scroll/destaque, ver
            # relatorios_robo.js.
            "processo_busca": processo,
            "erro": erro,
            "sucesso": sucesso,
        },
    )
    marcar_aba_vista(usuario.id, FERRAMENTA_SLUG, ABA_RELATORIOS_ROBO)

    return resposta


@router.get("/relatorios-robo/{job_id}/pdf")
def ver_pdf_relatorio_robo_route(job_id: int):
    """Ver docstring equivalente em relatorios_manuais.py — mesma lógica."""
    job = obter_job(job_id)

    if not job or not job.destino_pdf:
        raise HTTPException(status_code=404, detail="PDF de origem não encontrado.")

    caminho = Path(job.destino_pdf)

    if not caminho.exists():
        raise HTTPException(status_code=404, detail="PDF de origem não encontrado.")

    return FileResponse(caminho, media_type="application/pdf")


def _redirecionar(erro=None, sucesso=None):
    partes = []

    if erro:
        partes.append(f"erro={quote(erro)}")

    if sucesso:
        partes.append(f"sucesso={quote(sucesso)}")

    query = f"?{'&'.join(partes)}" if partes else ""

    return RedirectResponse(url=f"/extratus/relatorios-robo{query}", status_code=303)


@router.post("/relatorios-robo/{job_id}/excluir")
def excluir_relatorio_robo_route(job_id: int, usuario: Usuario = Depends(exigir_admin)):
    """Mesma regra do equivalente manual (relatorios_manuais.py): só
    admin da plataforma exclui de verdade."""
    if not excluir_job(job_id):
        return _redirecionar(erro="Esse relatório não existe mais.")

    return _redirecionar(sucesso="Relatório excluído permanentemente.")


@router.post("/relatorios-robo/baixar-lote")
def baixar_lote_relatorios_robo(ids: list[int] = Form(...)):
    """Baixa vários relatórios do Robô de uma vez, num .zip só — Henrique,
    2026-09-02: "Baixar todos" (respeitando os filtros já aplicados na
    tela) e "Baixar selecionados" (modo de seleção) caem os dois aqui,
    mesmo endpoint — só muda quais ids o JS manda (ver relatorios_robo.js).
    Job "erro" nunca tem `relatorio_path` (nunca gerou arquivo de
    verdade) — pulado tanto por essa checagem quanto pelo status em si,
    de propósito redundante: Henrique pediu explicitamente "revisão sim,
    erro não", então o status vira uma segunda trava explícita, não só
    uma consequência indireta de "sem arquivo"."""
    buffer = BytesIO()
    nomes_usados = set()

    with zipfile.ZipFile(buffer, "w", zipfile.ZIP_DEFLATED) as zip_arquivo:
        for job_id in ids:
            job = obter_job(job_id)

            if not job or job.status == "erro" or not job.relatorio_path:
                continue

            caminho = Path(job.relatorio_path)

            if not caminho.exists():
                continue

            nome_no_zip = caminho.name
            if nome_no_zip in nomes_usados:
                nome_no_zip = f"{caminho.stem}_{job.id}{caminho.suffix}"
            nomes_usados.add(nome_no_zip)

            zip_arquivo.write(caminho, arcname=nome_no_zip)

    if not nomes_usados:
        return _redirecionar(erro="Nenhum dos relatórios selecionados tem arquivo pra baixar.")

    nome_zip = f"relatorios_robo_{datetime.now().strftime('%Y%m%d_%H%M%S')}.zip"

    return Response(
        content=buffer.getvalue(),
        media_type="application/zip",
        headers={"Content-Disposition": f'attachment; filename="{nome_zip}"'},
    )


@router.post("/relatorios-robo/excluir-lote")
def excluir_lote_relatorios_robo(ids: list[int] = Form(...), usuario: Usuario = Depends(exigir_admin)):
    """Exclui vários relatórios do Robô de uma vez — mesma regra do
    excluir individual (só admin da plataforma), reaproveitando
    excluir_job por trás, um a um."""
    excluidos = 0

    for job_id in ids:
        if excluir_job(job_id):
            excluidos += 1

    if excluidos == 0:
        return _redirecionar(erro="Nenhum dos relatórios selecionados existe mais.")

    mensagem = (
        "1 relatório excluído permanentemente."
        if excluidos == 1
        else f"{excluidos} relatórios excluídos permanentemente."
    )
    return _redirecionar(sucesso=mensagem)


@router.post("/relatorios-robo/{job_id}/marcar-notificacao-resolvida")
def marcar_notificacao_resolvida_robo_route(
    job_id: int,
    usuario: Usuario = Depends(exigir_acesso_ferramenta("extratus")),
):
    """X do "sucesso" do Robô na aba "Ferramentas" do sino — diferente
    do equivalente manual (relatorios_manuais.py), não tem dono: qualquer
    um com acesso à ferramenta pode dispensar (Henrique, diretoria,
    2026-08-19). "Revisão" e "erro" do Robô não têm esse botão de
    propósito, mesma exigência de "não pode sumir sozinho" que erro já
    tinha."""
    if not marcar_notificacao_resolvida_robo(job_id):
        raise HTTPException(status_code=404, detail="Esse relatório não existe mais.")

    return {"ok": True}

File: app/ferramentas/extratus/web/static/relatorios_robo.js
(function () {
    "use strict";

    var campoBusca = document.querySelector(".campo-busca");
    var abas = document.querySelectorAll(".aba-relatorios-robo");
    var itens = document.querySelectorAll(".relatorio-item");
    var avisoVazio = document.querySelector(".filtro-vazio");
    var avisoSemSolicitacoes = document.getElementById("aviso-sem-solicitacoes-robo");
    var listaEl = document.getElementById("lista-relatorios-robo");
    var campoDataDe = document.getElementById("filtro-data-de");
    var campoDataAte = document.getElementById("filtro-data-ate");
    var campoSolicitante = document.getElementById("filtro-solicitante");

    if (!campoBusca || !itens.length) {
        return;
    }

    // Henrique, 2026-09-02: quem nunca solicitou nada ao Robô não tem
    // como aparecer pré-selecionado no dropdown "Solicitado por" (não
    // existe opção pra isso, de propósito — ver relatorios_robo.py). O
    // jeito de mesmo assim abrir a tela "filtrada em mim" é um flag à
    // parte, sem depender de nenhum valor de filtro de verdade: começa
    // forçando a lista inteira escondida (com a mensagem dedicada no
    // lugar), e qualquer interação real com QUALQUER filtro desliga isso
    // de vez — a pessoa nunca fica presa, só não vê o acervo inteiro sem
    // querer logo de cara.
    var filtroInicialForcado = !!(avisoSemSolicitacoes && avisoSemSolicitacoes.dataset.ativo === "true");

    // "Todos" é a aba padrão (Henrique, 2026-08-21, mudou de ideia em
    // relação à decisão de 2026-08-08 abaixo) — mostra do mais antigo
    // pro mais novo (column-reverse no CSS), diferente das outras 3
    // abas (Sucesso, Revisão, Erro — essas continuam mais novo primeiro,
    // "em ordem de grau de fudido").
    var statusAtivo = "todos";

    function aplicarFiltros() {
        if (filtroInicialForcado) {
            itens.forEach(function (item) { item.style.display = "none"; });
            if (avisoVazio) { avisoVazio.style.display = "none"; }
            // "block", não "" — esses <p> nascem com o atributo `hidden`
            // (não só sem classe de display nenhuma), e limpar o inline
            // style só devolve o controle pro `[hidden]` nativo do
            // navegador, que continua escondendo. Precisa de um valor
            // concreto pra vencer de vez (mesma pegadinha catalogada em
            // base.css, ex: .bandeja-apps[hidden]).
            if (avisoSemSolicitacoes) { avisoSemSolicitacoes.style.display = "block"; }
            return;
        }

        if (avisoSemSolicitacoes) {
            avisoSemSolicitacoes.style.display = "none";
        }

        var termo = campoBusca.value.trim().toLowerCase();
        var dataDe = campoDataDe ? campoDataDe.value : "";
        var dataAte = campoDataAte ? campoDataAte.value : "";
        var solicitanteId = campoSolicitante ? campoSolicitante.value : "";
        var visiveis = 0;

        itens.forEach(function (item) {
            var passaStatus = statusAtivo === "todos" || item.dataset.status === statusAtivo;
            var passaBusca = !termo || item.dataset.busca.indexOf(termo) !== -1;
            // Comparação de string funciona direto porque data-criado-em
            // e o <input type="date"> usam o mesmo formato ISO (AAAA-MM-DD).
            var passaData = (!dataDe || item.dataset.criadoEm >= dataDe)
                && (!dataAte || item.dataset.criadoEm <= dataAte);
            // Henrique, diretoria, 2026-08-27: filtro "Solicitado por" —
            // data-solicitante-id vem vazio quando o arquivo apareceu na
            // pasta por fora do upload da tela (ver job.solicitante_id,
            // carregado desde ChecagemFila.solicitante_id/
            // checagem_fila.registrar_pendente).
            var passaSolicitante = !solicitanteId || item.dataset.solicitanteId === solicitanteId;
            var mostrar = passaStatus && passaBusca && passaData && passaSolicitante;

            // Mesma pegadinha do [hidden] vs display de autor já
            // resolvida em relatorios_manuais.js — style.display direto
            // em vez do atributo.
            item.style.display = mostrar ? "" : "none";

            if (mostrar) {
                visiveis += 1;
            }
        });

        if (avisoVazio) {
            // Mesmo motivo do "block" acima — este <p> também nasce com
            // `hidden`.
            avisoVazio.style.display = visiveis === 0 ? "block" : "none";
        }
    }

    campoBusca.addEventListener("input", function () {
        filtroInicialForcado = false;
        aplicarFiltros();
    });

    // Henrique, 2026-08-26: o calendário de cada campo se ajusta pelo
    // que já foi escolhido no outro — não faz sentido "Até" permitir uma
    // data antes de "De" (nem o contrário). Quem for preenchido PRIMEIRO
    // vira o limite do outro; limpar o campo remove o limite de novo.
    if (campoDataDe) {
        campoDataDe.addEventListener("change", function () {
            filtroInicialForcado = false;
            if (campoDataAte) {
                campoDataAte.min = campoDataDe.value || "";
            }
            aplicarFiltros();
        });
    }
    if (campoDataAte) {
        campoDataAte.addEventListener("change", function () {
            filtroInicialForcado = false;
            if (campoDataDe) {
                campoDataDe.max = campoDataAte.value || "";
            }
            aplicarFiltros();
        });
    }

    if (campoSolicitante) {
        campoSolicitante.addEventListener("change", function () {
            filtroInicialForcado = false;
            aplicarFiltros();
        });
    }

    abas.forEach(function (aba) {
        aba.addEventListener("click", function () {
            filtroInicialForcado = false;
            abas.forEach(function (a) { a.classList.remove("aba-relatorios-robo-ativa"); });
            aba.classList.add("aba-relatorios-robo-ativa");
            statusAtivo = aba.dataset.status;

            if (listaEl) {
                listaEl.classList.toggle("lista-relatorios-invertida", statusAtivo === "todos");
            }

            aplicarFiltros();
        });
    });

    // Clicar no nome/meta do relatório baixa, igual à Relatórios manual
    // — .truncavel é tratado globalmente em base.js, não conflita.
    document.querySelectorAll(".relatorio-info-clicavel").forEach(function (bloco) {
        bloco.addEventListener("click", function () {
            window.location = bloco.dataset.download;
        });
    });

    // Nome do PDF de origem — ver comentário equivalente em
    // relatorios_manuais.js (Relatórios URGENTES) — mesma lógica.
    document.querySelectorAll(".link-pdf-original").forEach(function (link) {
        link.addEventListener("click", function (evento) {
            evento.stopPropagation();
        });
    });

    // Deep-link do botão "Ir ao relatório" (Conferências manuais,
    // web/routes/gerar_relatorio.py, ?processo=...) — pré-preenche a busca, troca
    // pra aba certa (o item pode estar em Sucesso/Revisão/Erro — mesmo
    // com "Todos" sendo o padrão, um deep-link deve destacar a aba
    // específica do item, não deixar em "Todos") e dá scroll+destaque no
    // item certo.
    var processoInicial = campoBusca.dataset.processoInicial;
    if (processoInicial) {
        var alvoPreCheck = document.querySelector('.relatorio-item[data-processo="' + CSS.escape(processoInicial) + '"]');
        if (alvoPreCheck) {
            var abaCerta = document.querySelector('.aba-relatorios-robo[data-status="' + alvoPreCheck.dataset.status + '"]');
            if (abaCerta) {
                abas.forEach(function (a) { a.classList.remove("aba-relatorios-robo-ativa"); });
                abaCerta.classList.add("aba-relatorios-robo-ativa");
                statusAtivo = alvoPreCheck.dataset.status;
            }
        }
        campoBusca.value = processoInicial;
        // Um deep-link sempre vence — nunca pode ficar escondido atrás do
        // filtro "só o que é meu" (padrão ou forçado), mesmo relatório
        // sendo de outra pessoa.
        filtroInicialForcado = false;
    }

    // Henrique, 2026-09-02: valor INICIAL do dropdown "Solicitado por" —
    // só pra quem tem pelo menos 1 solicitação de verdade (ver
    // sem_solicitacoes_proprias/filtroInicialForcado acima, o caso
    // oposto). Continua trocável livremente depois, igual ao "Solicitados
    // por mim" da tela manual.
    var solicitantePadrao = campoSolicitante ? campoSolicitante.dataset.padrao : "";
    if (solicitantePadrao && !processoInicial) {
        campoSolicitante.value = solicitantePadrao;
    }

    // "Todos" é o padrão agora — aplica a inversão de lista já na carga
    // inicial, não só quando alguém clica na aba (mesmo motivo do toggle
    // dentro do listener de clique acima).
    if (listaEl) {
        listaEl.classList.toggle("lista-relatorios-invertida", statusAtivo === "todos");
    }

    aplicarFiltros();

    if (processoInicial) {
        var alvo = document.querySelector('.relatorio-item[data-processo="' + CSS.escape(processoInicial) + '"]');
        if (alvo) {
            alvo.scrollIntoView({ behavior: "smooth", block: "center" });
            alvo.classList.add("relatorio-item-destacado");
            setTimeout(function () { alvo.classList.remove("relatorio-item-destacado"); }, 2400);
        }
    }

    // ---------------------------------------------------------------
    // "Baixar todos" + modo de seleção (checkboxes, baixar/excluir em
    // lote) — Henrique, 2026-09-02. Mesmo padrão de seleção já usado na
    // Fila do Robô (fila.js): botão "Selecionar" liga o modo, checkboxes
    // ficam escondidos até lá, "Selecionar todos" e o(s) botão(ões) de
    // ação habilitam/desabilitam junto com a contagem de marcados.
    // ---------------------------------------------------------------
    var botaoBaixarTodos = document.getElementById("botao-baixar-todos-robo");
    var botaoSelecionarRobo = document.getElementById("botao-selecionar-robo");
    var acoesSelecaoRobo = document.getElementById("relatorios-robo-acoes-selecao");
    var checkTodosRobo = document.getElementById("check-selecionar-todos-robo");
    var botaoBaixarSelecionados = document.getElementById("botao-baixar-selecionados-robo");
    var botaoExcluirSelecionados = document.getElementById("botao-excluir-selecionados-robo");
    var botaoCancelarSelecaoRobo = document.getElementById("botao-cancelar-selecao-robo");
    var formBaixarLoteRobo = document.getElementById("form-baixar-lote-robo");
    var formExcluirLoteRobo = document.getElementById("form-excluir-lote-robo");

    // Um relatório só entra em "baixar" quando tem arquivo de verdade —
    // Henrique, 2026-09-02: "o botão não pode incluir os com erro (revisão
    // sim, erro não)". `.botao-download:not(.botao-fantasma)` já reflete
    // isso (o template só desenha o ícone de baixar quando `job.
    // relatorio_path` existe, e "erro" nunca tem esse campo preenchido) —
    // checa os dois de propósito redundante, igual o backend faz.
    function relatorioTemArquivo(item) {
        return item.dataset.status !== "erro" && !!item.querySelector(".botao-download:not(.botao-fantasma)");
    }

    function submeterIdsEmLote(form, ids) {
        form.innerHTML = "";
        ids.forEach(function (id) {
            var campo = document.createElement("input");
            campo.type = "hidden";
            campo.name = "ids";
            campo.value = id;
            form.appendChild(campo);
        });
        form.submit();
    }

    // Henrique, 2026-09-02: "confirmação antes de baixar, pra evitar
    // download acidental de uma pasta GIGANTE" — mesmo modal já usado em
    // "Excluir selecionados", só que sem o vermelho de perigo (baixar não
    // é destrutivo, só pode ser sem querer/grande demais).
    function confirmarEBaixar(form, ids) {
        if (ids.length === 0) {
            return;
        }

        var mensagem = ids.length === 1
            ? "Baixar 1 relatório?"
            : "Baixar " + ids.length + " relatórios num arquivo .zip?";

        window.confirmarAcao(mensagem, function () {
            submeterIdsEmLote(form, ids);
        });
    }

    if (botaoBaixarTodos && formBaixarLoteRobo) {
        botaoBaixarTodos.addEventListener("click", function () {
            // "Todos" aqui respeita o que a tela está mostrando NA HORA —
            // filtros de data/status/solicitante/busca já aplicados, não
            // o histórico inteiro.
            var ids = Array.prototype.filter.call(itens, function (item) {
                return item.style.display !== "none" && relatorioTemArquivo(item);
            }).map(function (item) { return item.dataset.jobId; });

            confirmarEBaixar(formBaixarLoteRobo, ids);
        });
    }

    if (botaoSelecionarRobo && acoesSelecaoRobo) {
        function obterChecksRobo() {
            return listaEl.querySelectorAll(".relatorio-item-check");
        }

        function idsMarcados() {
            return Array.prototype.filter.call(obterChecksRobo(), function (c) { return c.checked; })
                .map(function (c) { return c.dataset.jobId; });
        }

        function atualizarBotoesSelecaoRobo() {
            var checks = obterChecksRobo();
            var algumMarcado = Array.prototype.some.call(checks, function (c) { return c.checked; });
            var todosMarcados = checks.length > 0 && Array.prototype.every.call(checks, function (c) { return c.checked; });

            if (botaoBaixarSelecionados) { botaoBaixarSelecionados.disabled = !algumMarcado; }
            if (botaoExcluirSelecionados) { botaoExcluirSelecionados.disabled = !algumMarcado; }
            if (checkTodosRobo) { checkTodosRobo.checked = todosMarcados; }
        }

        function entrarModoSelecaoRobo() {
            // Henrique, 2026-09-02 (3ª rodada): "Selecionar" some de vez
            // (não só esmaece) — igual "Baixar todos" — pra dar lugar aos
            // botões novos da barra de seleção, sem ficar um botão
            // desabilitado sobrando em cima.
            botaoSelecionarRobo.hidden = true;
            acoesSelecaoRobo.hidden = false;
            if (botaoBaixarTodos) { botaoBaixarTodos.hidden = true; }
            listaEl.querySelectorAll(".relatorio-item-checkbox").forEach(function (c) { c.hidden = false; });
            atualizarBotoesSelecaoRobo();
        }

        function sairModoSelecaoRobo() {
            botaoSelecionarRobo.hidden = false;
            acoesSelecaoRobo.hidden = true;
            if (botaoBaixarTodos) { botaoBaixarTodos.hidden = false; }
            listaEl.querySelectorAll(".relatorio-item-checkbox").forEach(function (c) { c.hidden = true; });
            obterChecksRobo().forEach(function (c) { c.checked = false; });
            if (checkTodosRobo) { checkTodosRobo.checked = false; }
        }

        botaoSelecionarRobo.addEventListener("click", entrarModoSelecaoRobo);
        botaoCancelarSelecaoRobo.addEventListener("click", sairModoSelecaoRobo);

        if (checkTodosRobo) {
            checkTodosRobo.addEventListener("change", function () {
                obterChecksRobo().forEach(function (c) { c.checked = checkTodosRobo.checked; });
                atualizarBotoesSelecaoRobo();
            });
        }

        // Delegação no container (não um listener por checkbox) — mesmo
        // motivo de fila.js: mais simples de manter, e já cobre qualquer
        // linha que passe a existir se essa lista um dia virar dinâmica.
        listaEl.addEventListener("change", function (evento) {
            if (evento.target.classList.contains("relatorio-item-check")) {
                atualizarBotoesSelecaoRobo();
            }
        });

        if (botaoBaixarSelecionados && formBaixarLoteRobo) {
            botaoBaixarSelecionados.addEventListener("click", function () {
                confirmarEBaixar(formBaixarLoteRobo, idsMarcados());
            });
        }

        if (botaoExcluirSelecionados && formExcluirLoteRobo) {
            botaoExcluirSelecionados.addEventListener("click", function () {
                var ids = idsMarcados();
                if (ids.length === 0) {
                    return;
                }

                var mensagem = ids.length === 1
                    ? "Excluir 1 relatório selecionado permanentemente? O arquivo e o PDF de origem serão apagados. Essa ação não pode ser desfeita."
                    : "Excluir " + ids.length + " relatórios selecionados permanentemente? Os arquivos e os PDFs de origem serão apagados. Essa ação não pode ser desfeita.";

                window.confirmarAcao(mensagem, function () {
                    submeterIdsEmLote(formExcluirLoteRobo, ids);
                }, true);
            });
        }
    }
})();
